### ISSUE TITLE
[Hint] Unnecessary vendor prefixed properties or rules

### ISSUE THREAD

**Issue by jaspreet57** [webhintio/hint#2845]
# 🚀 Hint request
Check the vendor prefixed properties used in CSS and report the properties which do not require to be prefixed for target browsers.

> It is good to remove unnecessary code. Though browsers will ignore such properties, but removing them will save the parsing time.

## Description
This hint will iterate through all prefixed rules, at-rules, properties and property values. If they are supported but are not required any more for given vendor and browser version, they will be reported by the hint.

**Note**: I think unsupported properties or rules should not be reported by this hint, as those are already being checked by `hint-compat-api/css`.

 

## 💪  Are you willing to work on this?
Yes, I wanted to start working on this hint as soon as it got discussed and approved.

## What should be tested exactly?
Hint should generate report for all the unnecessary vendor prefixed properties or rules in CSS.
For example if there are `-webkit-box-sizing` or `-webkit-max-content` in the CSS, It should be reported as unnecessary.

## Relevant links
I need a little bit discussion on this topic before adding useful resources here.

### 1.
To start with, I would like to understand one test case in `hint-compat-api/css` at below link:
https://github.com/webhintio/hint/blob/ae5d8be017289d71449a35d92d8f86c4fa61e1b4/packages/hint-compat-api/tests/css.ts#L123

It seems like if unprefixed version of particular prefixed rule is supported by the target browser, I can say that vendor prefix is not required any more and can be removed. Is it right?

### 2.
There is another observation I found that tools like [Autoprefixer](https://autoprefixer.github.io/) still add unnecessary vendor prefixes. Just test for `width: max-content`. `-webkit-` is not required here but `autoprefixer` add it.

So how could this hint be helpful in case if user is using auto prefixing tools? User wont be able to remove vendor prefixes and will have to ignore this hint.

**Comment by jaspreet57** [comment]
@antross Please help me in improving this hint description, and what else could be tested in this hint.

**Comment by antross** [comment]
> It seems like if unprefixed version of particular prefixed rule is supported by the target browser, I can say that vendor prefix is not required any more and can be removed. Is it right?

Yes, that's correct. To paraphrase a bit, if every target browser which supports a given vendor prefixed version of a particular feature (rule, property, etc.) also supports the unprefixed version, then the prefixed version can be removed.

> There is another observation I found that tools like Autoprefixer still add unnecessary vendor prefixes. Just test for width: max-content. -webkit- is not required here but autoprefixer add it.

Are you sure Autoprefixer is picking up the same target browsers as webhint in this case? If so, it may be due to a difference in data sources. Autoprefixer uses [Can I Use data](https://github.com/Fyrd/caniuse) to make its decisions whereas webhint uses [MDN data](https://github.com/mdn/browser-compat-data). We should help fix discrepancies we encounter by filing bugs against whichever data source is incorrect. 

An existing related issue is if webhint is run against raw source (e.g. via the VS Code extension) in a project where Autoprefixer is being used. In this case we wouldn't expect developers to include vendor-prefixed properties in their original source code, but webhint will report issues if they don't. We should consider updating `hint-compat-api/css` to account for this and not generate false positives. I'll file this as a separate bug.

> Please help me in improving this hint description, and what else could be tested in this hint.

I think you've got it mostly covered. What I'm not sure about yet is whether this makes more sense as a separate hint or as an enhancement to `hint-compat-api/css`.

**Comment by jaspreet57** [comment]
> I think you've got it mostly covered. What I'm not sure about yet is whether this makes more sense as a separate hint or as an enhancement to hint-compat-api/css.

I am thinking same. But `hint-compat-api/css` seems to test if property is supported or not. Where prefixed property might still be supported but not required.

If we want to include it in `hint-compat-api/css`, we need to update its description that it also checks for unnecessary vendor prefixes.

Thought of others might help more.

> Are you sure Autoprefixer is picking up the same target browsers as webhint in this case?

I was actually testing Autoprefixer for 'last 1 version'.

![autoprefix-last-1-version](https://user-images.githubusercontent.com/10244949/63530563-52ca6080-c524-11e9-8d6a-8227fb8c62cb.png)

Where `last 1 version` include many targets as [given here](https://browserl.ist/?q=last+1+version). Out of which it seems `iOS Safari12.2-12.38.4` still require `-webkit-` as indicated by [canIUse](https://caniuse.com/#search=max-content).

So 'Can I use' data is correct. I just need to specify only chrome version to test it.

This is fine, but I am curious that how user specify the target version to Autoprefixer? Is it like `last n version` or specific browser with version.

**Comment by antross** [comment]
> If we want to include it in hint-compat-api/css, we need to update its description that it also checks for unnecessary vendor prefixes.

Yep, that's fine.

> This is fine, but I am curious that how user specify the target version to Autoprefixer? Is it like last n version or specific browser with version.

Lots of different options for queries, see https://github.com/browserslist/browserslist#queries. Both Autoprefixer and webhint use `browserslist` to to select the set of browsers so the same query should produce the same result between both of them.

**Comment by jaspreet57** [comment]
> > If we want to include it in hint-compat-api/css, we need to update its description that it also checks for unnecessary vendor prefixes.
> 
> Yep, that's fine.
> 
> > This is fine, but I am curious that how user specify the target version to Autoprefixer? Is it like last n version or specific browser with version.
> 
> Lots of different options for queries, see https://github.com/browserslist/browserslist#queries. Both Autoprefixer and webhint use `browserslist` to to select the set of browsers so the same query should produce the same result between both of them.

Hmm cool. Thanks for the info !
